SMTChecker: Introduce first draft of Z3CHCSmtlib2Interface

blishko opened the pull request:
This commit introduces the interface class without actually using it. The actual switch will be done later, after all things are set up.
This was separated from #15252.

Commit:
SMTChecker: Introduce first draft of Z3CHCSmtlib2Interface

This commit introduces the interface class without actually using it.
The actual switch will be done later, after all things are set up.

r0qs commented on libsolidity/formal/Z3CHCSmtLib2Interface.cpp:
```diff
@@ -0,0 +1,214 @@
+		setupSmtCallback(true);
+		if (!boost::starts_with(response, "unsat"))
+			return {CheckResult::SATISFIABLE, Expression(true), {}};
+		return {CheckResult::SATISFIABLE, Expression(true), graphFromZ3Answer(response)};
```
```suggestion
		return {CheckResult::UNSATISFIABLE, Expression(true), graphFromZ3Answer(response)};
```

Shouldn't this be unsatisfiable?

blishko replied:
This is a bit confusing but let me explain:
When you use Z3 as a library, it uses the original Datalog semantics where SAT means UNSAFE and UNSAT means SAFE.
SMTChecker followed this convention.
However, when you use SMT-LIB format with `HORN` logic, the semantics is flipped. SAT means SAFE and UNSAT means UNSAFE.
That's why we need to flip the answer from the solver when we propagate it.
See also `CHCSmtLib2Interface::query` where the same flipping is implemented as well.

nikola-matic replied:
This is indeed confusing, and just a little bit insane, so a comment outlying this would be very much appreciated :)

blishko replied:
It is very confusing if you are not familiar with CHCs.
The thing is that it flips the question you are asking.
In the usual SMT-based model checking you are asking the solver "Is there a counterexample path?"
With Horn solvers, you are asking a different question, "Are there safe inductive invariants?"
So the meaning of the YES/NO answer is flipped.

I'll add a comment in the code.

blishko replied:
Comment added here and also to `CHCSmtLib2Interface`

r0qs commented on libsolidity/formal/Z3CHCSmtLib2Interface.cpp:
```diff
@@ -0,0 +1,214 @@
+	if (boost::starts_with(response, "sat"))
+	{
+		auto maybeInvariants = invariantsFromSolverResponse(response);
+		return {CheckResult::UNSATISFIABLE, maybeInvariants.value_or(Expression(true)), {}};
```
```suggestion
		return {CheckResult::SATISFIABLE, maybeInvariants.value_or(Expression(true)), {}};
```
And shouldn't here be satisfiable? Maybe I'm missing something.

r0qs commented on libsolidity/formal/Z3CHCSmtLib2Interface.cpp:
```diff
@@ -0,0 +1,214 @@
+		response = querySolver(query);
+		setupSmtCallback(true);
+		if (!boost::starts_with(response, "unsat"))
+			return {CheckResult::SATISFIABLE, Expression(true), {}};
```
If it is not `unsat` couldn't it also be `UNKNOWN` or `ERROR` here?

blishko replied:
If solver for some reason would not return `unsat` (which is what we expect), then we ignore this second call and just return what we would have returned if we would not want to compute counterexample.
See my response below for the SAT/UNSAT confusion.

r0qs commented on libsolidity/formal/Z3CHCSmtLib2Interface.cpp:
```diff
@@ -0,0 +1,214 @@
+			continue;
+
+		auto const& args = asSubExpressions(command);
+		auto const& head = args[0];
```
Shouldn't we assert that `args.size() > 0` before this line?

blishko replied:
Assert added!

nikola-matic commented on libsolidity/formal/Z3CHCSmtLib2Interface.h:
```diff
@@ -0,0 +1,49 @@
+	CHCSolverInterface::CexGraph graphFromZ3Answer(std::string const& _proof) const;
+	static CHCSolverInterface::CexGraph graphFromSMTLib2Expression(
+		smtutil::SMTLib2Expression const& _proof,
+		ScopedParser & context
```
```suggestion
		ScopedParser& _context
```

blishko replied:
Fixed.

nikola-matic commented on libsolidity/formal/Z3CHCSmtLib2Interface.cpp:
```diff
@@ -0,0 +1,215 @@
+	{
+		return {};
+	}
+	solAssert(parser.isEOF());
```
```suggestion
	solAssert(!parser.isEOF());
```

If it's EOF, we're not longer in the function, no?

blishko replied:
No, the original version is correct.
Parser parses the string representation and creates an `SMTLib2Expression` out of that.
Here we just assert the whole input has been parsed successfully.
The rest of the method then processes the parsed expression.

nikola-matic commented on libsolidity/formal/Z3CHCSmtLib2Interface.cpp:
```diff
@@ -0,0 +1,215 @@
+		if (!isAtom(head))
+			continue;
+
+		if (asAtom(head) == "declare-fun")
```
Can you add an example command for each of the `head`s? I believe this would be helpful when reviewing, especially for us non-smt people.

`// e.g. declare-fun something1 something2 something3`

blishko replied:
Added hopefully useful comment.

Regarding the `proof` expression, I added a comment there as well, but it is hard to give an example, because it really contains the whole proof.

nikola-matic commented on libsolidity/formal/Z3CHCSmtLib2Interface.cpp:
```diff
@@ -0,0 +1,215 @@
+			solAssert(!isAtom(domainSorts));
+			expressionParser.addVariableDeclaration(asAtom(name), expressionParser.toSort(codomainSort));
+		}
+		else if (asAtom(head) == "proof")
```
In this case, do we make any use of subexpressions (`args`), and if no, can this be asserted in some way? E.g. if `proof` is standalone, then `solAssert(args.size() == 1)`?

blishko replied:
I have `asserts` related to subexpressions in the method `graphFromSMTLib2Expression` which receives the whole `proof` expression (`command` variable).

nikola-matic commented on libsolidity/formal/Z3CHCSmtLib2Interface.cpp:
```diff
@@ -0,0 +1,215 @@
+
+		if (asAtom(head) == "declare-fun")
+		{
+			solAssert(args.size() == 4);
```
On the topic of asserts - these should be `smtAssert(...)`.

blishko replied:
Agreed. I think I have that in one of subsequent commits on my other branch.

nikola-matic commented on libsolidity/formal/Z3CHCSmtLib2Interface.cpp:
```diff
@@ -0,0 +1,215 @@
+
+	auto const* root = proofStack.top();
+	auto const& derivedRootFact = fact(*root);
+	visitedIds.insert({root, nextId++});
```
```suggestion
	visitedIds.emplace(root, nextId++);
```

blishko replied:
Fixed.

nikola-matic commented on libsolidity/formal/Z3CHCSmtLib2Interface.cpp:
```diff
@@ -0,0 +1,215 @@
+				auto const* child = &args[i];
+				if (!visitedIds.count(child))
+				{
+					visitedIds.insert({child, nextId++});
```
```suggestion
					visitedIds.emplace(child, nextId++);
```

blishko replied:
Fixed.

nikola-matic commented:
Seeing as this is an unused interface, I expect you're going to be making some tweaks in subsequent PR regardless, so please take a glance at the comments, and then we can merge.

blishko commented:
@nikola-matic, I think I addressed all your comments, except the `solAssert -> smtAssert`. I believe I have that prepared somewhere down the pipeline. If you are OK with it, I would address this one later.

nikola-matic commented:
> @nikola-matic, I think I addressed all your comments, except the `solAssert -> smtAssert`. I believe I have that prepared somewhere down the pipeline. If you are OK with it, I would address this one later.

Yup, that's fine.
